### PR TITLE
#37 #38 implement most Zone and all Zone Plan endpoints

### PR DESCRIPTION
**Pull request by jamesog**
As mentioned in #37 I haven't added CreateZone or EditZone yet, but wanted to put the rest up for review.

### CODE REVIEW

File: cloudflare.go [hunk 1]
```diff
@@ -116,12 +116,17 @@ func (api *API) makeRequest(method, uri string, params interface{}) ([]byte, err
 	return resBody, nil
 }
 
+type ResponseInfo struct {
```
Comment (elithrar) [inline cloudflare.go]: godoc comment?

File: cloudflare.go [hunk 2]
```diff
@@ -116,12 +116,17 @@ func (api *API) makeRequest(method, uri string, params interface{}) ([]byte, err
...
+
 // Response is a template.  There will also be a result struct.  There will be a
 // unique response type for each response, which will include this type.
 type Response struct {
-	Success  bool     `json:"success"`
-	Errors   []string `json:"errors"`
-	Messages []string `json:"messages"`
+	Success  bool           `json:"success"`
+	Errors   []ResponseInfo `json:"errors"`
+	Messages []ResponseInfo `json:"messages"`
```
Comment (elithrar) [inline cloudflare.go]: Is `messages` always a `[]string` or does it mirror `errors`? https://api.cloudflare.com/#responses

File: zone.go [hunk 3]
```diff
@@ -141,14 +234,56 @@ func (api *API) PurgeCache(zoneID string, pcr PurgeCacheRequest) (PurgeCacheResp
...
+		return ZoneID{}, errors.Wrap(err, errUnmarshalError)
+	}
+	return r.Result, nil
+}
+
+// AvailablePlans returns information about all plans available to the specified zone.
+// API reference:
+//  https://api.cloudflare.com/#zone-plan-available-plans
+//  GET /zones/:zone_identifier/available_plans
+func (api *API) AvailablePlans(zoneID string) ([]ZonePlan, error) {
```
Comment (elithrar) [inline zone.go]: Or `AvailableZonePlans`?

File: zone.go [hunk 4]
```diff
@@ -141,14 +234,56 @@ func (api *API) PurgeCache(zoneID string, pcr PurgeCacheRequest) (PurgeCacheResp
...
 }
 
-// Zone Plan
-// https://api.cloudflare.com/#zone-plan-available-plans
-// https://api.cloudflare.com/#zone-plan-plan-details
+// PlanDetails returns information about a zone plan.
+// API reference:
+//  https://api.cloudflare.com/#zone-plan-plan-details
+//  GET /zones/:zone_identifier/available_plans/:identifier
+func (api *API) PlanDetails(zoneID, planID string) (ZonePlan, error) {
```
Comment (elithrar) [inline zone.go]: Or `ZonePlanDetails`?

**Comment by elithrar** [comment]
Looking pretty good to me! Also addressed your questions in #37.

File: cloudflare.go (hunk 1 above)
Comment (jamesog) [inline cloudflare.go]: Whoops, will fix that. Thanks.

File: cloudflare.go (hunk 2 above)
Comment (jamesog) [inline cloudflare.go]: `messages` and `errors` seem to be the same.

File: zone.go (hunk 3 above)
Comment (jamesog) [inline zone.go]: That does give a bit more context. I'll change that. Same for the next one. Thanks.

File: zone.go [hunk 5]
```diff
@@ -67,31 +174,17 @@ func (api *API) ListZones(z ...string) ([]Zone, error) {
 // API reference:
 // 	https://api.cloudflare.com/#zone-zone-details
```
Comment (elithrar) [inline zone.go]: Tab?

File: zone.go [hunk 6]
```diff
@@ -15,7 +108,21 @@ func (api *API) CreateZone(z Zone) {
 	// res, err := api.makeRequest("POST", "/zones", z)
 }
 
-// ListZones liist zones on an account. Optionally takes a list of zone names
+// ZoneActivationCheck initiates another zone activation check for newly-created zones.
+// API reference:
+//   https://api.cloudflare.com/#zone-initiate-another-zone-activation-check
```
Comment (elithrar) [inline zone.go]: 2 spaces? (gofmt oddity?)

File: zone.go (hunk 5 above)
Comment (jamesog) [inline zone.go]: Oh weird, not sure how that happened.

File: zone.go (hunk 6 above)
Comment (jamesog) [inline zone.go]: A me oddity :-) I was originally adding multiple spaces to have it mono-spaced in godoc but I see you made that a single space after your golint commit. Gofmt doesn't seem to care about spaces in comments so didn't change it. I'll update to match yours.

**Comment by elithrar** [comment]
This just needs a rebase & some conflicts resolved.

File: cmd/flarectl/flarectl.go [hunk 7]
```diff
@@ -504,6 +552,37 @@ func main() {
 					Usage:   "List all zones on an account",
 				},
 				{
+					Name:    "create",
+					Aliases: []string{"c"},
+					Action:  zoneCreate,
+					Usage:   "List all zones on an account",
```
Comment (jamesog) [inline cmd/flarectl/flarectl.go]: Well that's clearly not right. Will fix.

**Comment by jamesog** [comment]
Ah, still need to implement `EditZone`. Will get that done before merging.

**Comment by jamesog** [comment]
I think this is done now. Just one issue - should be fully-functional, but one issue with `ZoneSetPlan` as noted in the commit message and in the code with a `FIXME`.

**Comment by jamesog** [comment]
(I've also got a branch ready to address #39 but requires this one to be merged first.)

File: zone.go [hunk 8]
```diff
@@ -7,23 +7,157 @@ import (
...
+// ZoneActivationCheck initiates another zone activation check for newly-created zones.
+//
+// API reference: https://api.cloudflare.com/#zone-initiate-another-zone-activation-check
+func (api *API) ZoneActivationCheck(zoneID string) (Response, error) {
+	res, err := api.makeRequest("PUT", "/zones/"+zoneID+"/activation_check", nil)
+	if err != nil {
+		return Response{}, errors.Wrap(err, errMakeRequestError)
+	}
+	var r Response
+	err = json.Unmarshal(res, &r)
```
Comment (nishanths) [inline zone.go]: Check `err != nil` after `Unmarshal`

File: zone.go [hunk 9]
```diff
@@ -7,23 +7,157 @@ import (
...
+		newzone.Organization = &org
+	}
+
+	res, err := api.makeRequest("POST", "/zones", newzone)
+	if err != nil {
+		return Zone{}, errors.Wrap(err, errMakeRequestError)
+	}
+
+	var r ZoneResponse
+	err = json.Unmarshal(res, &r)
```
Comment (nishanths) [inline zone.go]: Check `err != nil` after `Unmarshal`

File: zone.go (hunk 8 above)
Comment (jamesog) [inline zone.go]: Good catch, thanks. Must have missed that after debugging.

**Comment by jamesog** [comment]
Any more comments or can we merge this one in?

**Comment by elithrar** [comment]
LGTM!

**Pull request merged**